app/scripts/lib/util.ts

```diff
@@ -480,3 +481,138 @@ export function getConversionRatesForNativeAsset({
 
   return conversionRateResult;
 }
+
+// Cache for known domains
+let knownDomainsSet: Set<string> | null = null;
+let initPromise: Promise<void> | null = null;
+
+/**
+ * Initialize the set of known domains from the chains list
+ */
+export async function initializeRpcProviderDomains(): Promise<void> {
+  if (initPromise) {
+    return initPromise;
+  }
+
+  initPromise = (async () => {
+    try {
+      const chainsList = await getSafeChainsList();
+      knownDomainsSet = new Set<string>();
+
+      for (const chain of chainsList) {
+        if (chain.rpc && Array.isArray(chain.rpc)) {
+          for (const rpcUrl of chain.rpc) {
+            try {
+              // Handle placeholder variables like ${INFURA_API_KEY}
+              const cleanedUrl = rpcUrl.replace(/\$\{[^}]+\}/gu, 'API_KEY');
+              const url = new URL(cleanedUrl);
+              knownDomainsSet.add(url.hostname);
+            } catch (e) {
+              // Skip invalid URLs
+              continue;
+            }
+          }
+        }
+      }
+    } catch (error) {
+      console.error('Error initializing known domains:', error);
+      knownDomainsSet = new Set<string>();
+    }
+  })();
+
+  return initPromise;
+}
+
+/**
+ * Check if a domain is in the known domains list
+ *
+ * @param domain - The domain to check
+ */
+export function isKnownDomain(domain: string): boolean {
+  return knownDomainsSet?.has(domain) ?? false;
+}
+
+/**
+ * Extracts the domain from an RPC endpoint URL with privacy considerations
+ *
+ * @param rpcUrl - The RPC endpoint URL
+ * @param knownDomainsForTesting - Optional Set of known domains for testing purposes
+ * @returns The domain for known providers, 'private' for private/custom networks, or 'invalid' for invalid URLs
+ */
+export function extractRpcDomain(
+  rpcUrl: string,
+  knownDomainsForTesting?: Set<string>,
+): string {
+  if (!rpcUrl) {
+    return 'invalid';
+  }
+
+  const trimmedRpcUrl = rpcUrl.trim();
+
+  try {
+    let url;
+    try {
+      url = new URL(trimmedRpcUrl);
+    } catch (e) {
+      if (
+        !trimmedRpcUrl.startsWith('http://') &&
+        !trimmedRpcUrl.startsWith('https://')
+      ) {
+        try {
+          url = new URL(`https://${trimmedRpcUrl}`);
+        } catch (e2) {
+          return 'invalid';
+        }
```

Comment:
nit: I don't think this error handling logic will ever be used in practice, since our interface and API will always require a complete and valid URL. If it wasn't valid, this code could never run anyway, since it'd fail earlier. I think you can just return `invalid` right away.

Reply:
yes I was thinking the same after I committed it. Came back and tested it out. Your right.

Reply:
I've also removed some tests that made this assumption and knowing the API will always require a complete and valid URL, I've tried to remove anything that is uneeeded